pkg/controllers/v1/mpi_job_controller.go

```diff
@@ -834,6 +834,14 @@ func (c *MPIJobController) deleteWorkerPods(mpiJob *kubeflow.MPIJob) error {
 			c.recorder.Event(mpiJob, corev1.EventTypeWarning, ErrResourceExists, msg)
 			return fmt.Errorf(msg)
 		}
+		// If the worker pod is not running and cleanupPolicy is
+		// set to CleanPodPolicyRunning, keep the pod.
+		// Note that pending pod should still be removed under this
+		// situation, since it may turn to running in the future.
+		if *mpiJob.Spec.CleanPodPolicy == common.CleanPodPolicyRunning && !isPodRunning(pod) && !isPodPending(pod) {
```

Comment:
Make sense, but maybe you need to also change the logic here https://github.com/kubeflow/common/blob/31a6ec10d140a3c8cb12a863dcc1363f46ad9998/pkg/controller.v1/common/job.go#L31 if we decide to remove pending pods.

/cc @terrytangyuan @carmark @Jeffwan